azkaban-common/src/main/java/azkaban/utils/DependencyTransferManager.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2019 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.utils;
+
+import azkaban.spi.DependencyFile;
+import azkaban.spi.Storage;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.io.IOUtils;
+
+import static azkaban.utils.ThinArchiveUtils.*;
+
+/**
+ * handles downloading of dependencies. Used during the thin archive upload process, and upon starting the execution
+ * of a flow defined in a thin archive (to download necessary dependencies). Provides a thin layer of retry logic,
+ * checksum validation and parallelism on top of the base Storage::getDependency().
+ */
+@Singleton
+public class DependencyTransferManager {
+  public static final int MAX_DEPENDENCY_DOWNLOAD_TRIES = 2;
```

Comment:
Is it better to make it configurable instead of hardcoded to some value? Also would you consider some backoff time before the next retry?

Reply:
Both good ideas. `MAX_DEPENDENCY_DOWNLOAD_TRIES` is now configurable with the `azkaban.dependency.max.download.tries` property and an exponential delay is added for tries up to the maximum value.